docs/proving-contract-properties.md

```diff
@@ -0,0 +1,100 @@
+5) *formally verifiable*, i.e., some propositions about contract behavior can be proven as 
+theorems
+
+All these properties contribute to reliability, financial safety, and long-term survivability
+of the blockchain platform. Ergo contracts are written in ErgoScript language 
+(which is based on Scala syntax and semantics), and here we describe an environment for certified development of Ergo contracts.
+
+## Scope
+
+We aim to integrate formal contract verification into Ergo contract development toolbox
+without sacrificing other important properties, which essential for the development of 
+real-world contracts and mentioned in [Introduction](#introduction).
+
+#### Goals
+ - G1: allow usage of the same language (which is a subset of Scala) both for writing 
+ contracts and for formulation and verification of pre(post)-conditions 
+ - G2: allow the definition of reusable lemmas and modular proofs
+ - G3: allow verification for all Ergo contracts (both [simple](Crowdfunding) and 
+ [complex](ICO) contracts should be supported)
+ - G4: allow both verification and execution of the same source code without any modifications
+ - G5: allow usage of pre(post)-conditions for automatic property testing
+ - G6: allow a certified definition of ErgoScript primitive operations
+ - G7: allow automatic property testing of runtime implementation of ErgoScript primitives 
+ against their certified definitions
+ 
+#### Non-Goals
+ - development of new verification techniques
+ - development of a verification approach disconnected from the rest of contract development
+   toolchain.
+ - demonstration of verification on toy examples 
+
+## Ergo Contract Development Environment
+
+We have chosen the Scala language and tool ecosystem as the development environment for both 
+contract development and also for implementation of the ErgoScript language.
+
+Moreover, we designed ErgoScript to be a subset of Scala language, inheriting its syntax, 
+types and semantics of the selected Scala subset. We call this subset ErgoScala, keeping in mind that it is equivalent to ErgoScript and is also a valid Scala.
+
+Such defined ErgoScala allows us to meet the [goals](#goals) by:
+1) writing, executing, testing and debugging Ergo contracts directly in Scala using existing 
+tools, libraries and IDE support
+2) implementing Scala macros to compile ErgoScala fragments (contracts) into [ErgoTree](ref) 
+the intermediate representation of Ergo contracts suitable for storage in Ergo blockchain. (G4)
+3) using [Stainless](https://github.com/epfl-lara/stainless) verification framework to specify pre(post)-conditions for Ergo contracts and formally prove them using auxiliary lemmas. (PureScala is a superset of ErgoScala, 
+thus any ErgoScala fragment is verifiable as long as it is verifiable as PureScala fragment
+in Stainless) (G1, G2, G3)
+4) check the declared pre(post)-conditions using ScalaCheck property testing (G5)
```

Comment:
@aslesarenko How to check a pre-condition itself (e.g. `HEIGHT > 0`)? Do you mean to check on the contract that the post-conditions hold?

Reply:
@greenhat Yes, check post-conditions given pre-conditions. The pre-conditions should be checked as part of contract usage (i.e as part of transaction creation)

Reply:
@aslesarenko It's Stainless job to test if post-conditions hold. What's the point to test them with property testing(ScalaCheck)?